feat(topology/metric_space/antilipschitz): define `antilipschitz_with`

urkud opened the pull request:
Also make some proofs use facts about `antilipschitz_with`.

Depends on #2074

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [ ] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

urkud committed:
chore(data/real/ennreal): weaker assumptions in `sub_mul`, add `coe_inv_le`

urkud committed:
feat(topology/metric_space/antilipschitz): define `antilipschitz_with`

Also make some proofs use facts about `antilipschitz_with`.

urkud commented on src/topology/metric_space/lipschitz.lean:
```diff
@@ -50,6 +50,15 @@ variables [emetric_space α] [emetric_space β] [emetric_space γ] {K : ℝ≥0}
 lemma edist_le (h : lipschitz_with K f) (x y : α) : edist (f x) (f y) ≤ K * edist x y := h x y
 
+lemma edist_ge (h : lipschitz_with K f) (x y : α) :
+  (K⁻¹ : ennreal) * edist (f x) (f y) ≤ edist x y :=
```
I don't like this pair of names (`edist_le`, `edist_ge`) but failed to come up with another pair of short names.

urkud commented on src/topology/metric_space/antilipschitz.lean:
```diff
@@ -0,0 +1,105 @@
+  exact this.elim (λ hK', absurd hK' (ne_of_gt hK)) id
+end
+
+lemma edist_le (hf : antilipschitz_with K f) (hK : 0 < K) (x y) :
```
For `hf : lipschitz_with K f`, `hf.edist_le` is the defining property while `edist_ge` is a version with `K` moved to another side of the inequality. Here `edist_le` is a version with `K` moved to the right. Suggestions of better names are welcome.

sgouezel replied:
You could add a `mul` on the side where there is the constant, and use `edist` to refer to the edistance of the images. So, for a Lipschitz function, the inequality `edist (f x) (f y) <= K * edist x y` would be `edist_le_mul`, while your inequality here would be `le_mul_edist`, and the defining inequality for antilipschitz functions would be `mul_le_edist`, and so on.

urkud commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -233,11 +238,14 @@ lemma op_norm_le_bound {M : ℝ} (hMp: 0 ≤ M) (hM : ∀ x, ∥f x∥ ≤ M * 
-  ⟨add_nonneg (op_norm_nonneg _) (op_norm_nonneg _), λ x, by { rw add_mul,
-    exact norm_add_le_of_le (le_op_norm _ _) (le_op_norm _ _) }⟩
+show ∥f + g∥ ≤ (coe : nnreal → ℝ) (⟨_, f.op_norm_nonneg⟩ + ⟨_, g.op_norm_nonneg⟩),
+from op_norm_le_of_lipschitz (f.lipschitz.add g.lipschitz)
```
Not sure if this proof is better than the original one. Should I revert this change?

sgouezel replied:
I think your proof is fine.

sgouezel commented on src/analysis/normed_space/basic.lean:
```diff
@@ -132,6 +136,11 @@ lemma dist_sub_sub_le_of_le {g₁ g₂ h₁ h₂ : α} {d₁ d₂ : ℝ}
   dist (g₁ - g₂) (h₁ - h₂) ≤ d₁ + d₂ :=
 le_trans (dist_sub_sub_le g₁ g₂ h₁ h₂) (add_le_add H₁ H₂)
 
+lemma dist_add_add_ge (g₁ g₂ h₁ h₂ : α) :
```
I would use a longer name, like `abs_dist_sub_le_dist_add_add`.

urkud committed:
Rename inequalities, move `K` to the other side

This way it's easier to glue it with the rest of the library, and
we can avoid assuming `0 < K` in many lemmas.

urkud committed:
Merge branch 'master' into antilipschitz

mergify[bot] committed:
Merge branch 'master' into antilipschitz
